ocaml/xapi/xapi_host.ml

```diff
@@ -2671,44 +2671,165 @@ let with_temp_file_contents ~contents f =
+            if contents <> "" then (
+              with_temp_file_contents ~contents
+                (Tar_unix.Archive.extract extract_certificate_file) ;
+              debug "UEFI tar file extracted to temporary directory"
+            )
+        (* No UEFI tar file. *)
+        | Error _ ->
+            debug
+              "UEFI tar file was not extracted: it was not base64-encoded \
+               correctly"
+    )
+
+let write_uefi_certificates_to_disk ~__context ~host =
+  let ( // ) = Filename.concat in
+  let with_valid_symlink from_path to_path fn =
+    debug "override_uefi_certs: with_valid_symlink %s->%s" from_path to_path ;
+    let is_valid_symlink from_path to_path =
+      if Sys.file_exists from_path then
+        Unix.(
+          match (lstat from_path).st_kind with
+          | S_LNK ->
+              readlink from_path = to_path
+          | _ ->
+              false
         )
-    (* No UEFI tar file. *)
-    | Error _ ->
-        debug
-          "UEFI tar file was not extracted: it was not base64-encoded correctly"
+      else
+        false
+    in
+    if not (is_valid_symlink from_path to_path) then (
+      try
+        Helpers.rmtree from_path ;
+        Unix.symlink to_path from_path
+      with e ->
+        let msg = Printf.sprintf "Failed to symlink %s->%s" from_path to_path in
+        error "%s: %s" msg (Printexc.to_string e) ;
+        raise Api_errors.(Server_error (internal_error, [msg]))
+    ) ;
+    fn from_path
+  in
+  let with_empty_dir path fn =
+    debug "override_uefi_certs: with_empty_dir %s" path ;
+    if Sys.file_exists path then (
+      Unix.(
+        match (lstat path).st_kind with
+        | S_DIR ->
+            Sys.readdir path
+            |> Array.iter (fun file -> Helpers.rmtree (path // file))
+        | _ ->
+            Unix.unlink path ;
+            Unixext.mkdir_rec path 0o755
+      )
+    ) else
+      Unixext.mkdir_rec path 0o755 ;
+    fn path
+  in
+  let check_valid_uefi_certs_in path =
+    (* check expected uefi certificates are present *)
+    let rec with_uefi_certs_in_disk path =
+      if Sys.file_exists path then
+        Unix.(
+          match (lstat path).st_kind with
+          | S_DIR ->
+              Sys.readdir path |> Array.to_list
+          | S_LNK ->
+              Unix.readlink path |> with_uefi_certs_in_disk
+          | _ ->
+              []
+        )
+      else
+        []
+    in
+    let uefi_certs_in_disk = with_uefi_certs_in_disk path in
+    ["KEK.auth"; "db.auth"]
+    |> List.iter (fun cert ->
+           let log_of found =
+             (if found then debug else error)
+               "check_valid_uefi_certs: %s %s in %s"
+               (if found then "found" else "missing")
+               cert path
+           in
+           log_of (List.mem cert uefi_certs_in_disk)
+       )
+  in
+  match !Xapi_globs.override_uefi_certs with
+  | false ->
+      with_valid_symlink !Xapi_globs.varstore_dir !Xapi_globs.default_auth_dir
```

Comment:
as a curiosity, you might like this way of expressing the with_ functions:

```ocaml
    let ( let@ ) f x = f x
    let@ path = with_valid_symlink !Xapi_globs.varstore_dir !Xapi_globs.default_auth_dir in
    check_valid_uefi_certs_in path
```

similarly in the other case:
```ocaml
      let@ path = with_empty_dir !Xapi_globs.varstore_dir in
      let pool_uefi_certs =
        Db.Pool.get_uefi_certificates ~__context
          ~self:(Helpers.get_pool ~__context)
      in
      really_write_uefi_certificates_to_disk ~__context ~host
        ~value:pool_uefi_certs ;
      check_valid_uefi_certs_in path
```

Reply:
thanks, I like this